### PR TITLE
Run VSCode extension's tests on top of che-theia

### PR DESCRIPTION
**Pull request by svor**
### What does this PR do?
This PR contains an action **VSCode Extension's Tests** to run tests from VSCode extension on top of Che-Theia if it was changed in `che-theia-plugins.yaml`.

VSCode Extension's Tests workflow:
https://github.com/svor/che-plugin-registry/actions/workflows/vscode-extension-tests-check.yaml
screenshot-github com-2021 03 01-14_50_59
1. The first step is **Checkout**. It clones the repository and makes checkout to the working branch.

2. Step 2 is **Get changes**. This step executes `./ci/get-changed-plugin.sh`. In this script we check what exactly was changed in `che-theia-plugins.yaml`. If this file doesn't contain changes on lines with key `revision` - it means that none extension is going to be updated or added and in this way we set `EXTENSION_REPO` environment variable as empty and all next action's steps will be skipped (action should be finished). In another way, when some value of revision was changed, we find all information about the repository of changed extension (url and a revision).

3. Step 3 is **Build image**. It builds `che-plugin-registry` image with all updates. It is needed to run the workspace with updated/added plugin.

4. Step 4 is **Start minikube**. Needs to setup minikube

5. Step 5 is **Load image**. It loads che-plugin registry image into docker registry

6. Step 6 is **Deploy Eclipse Che**. This step uses `che-incubator/che-deploy-action@next` action to deploy Che with custom che-plugin-registry image into minikube.

7. Step 7 is **Run VSCode extension's tests**. This step executes `./ci/run-extension-tests.sh`. The script has a few steps:
	-  First of all we need to install all required dependencies into the runner (selenium, geckodriver)
	-  By using `url` and `revision` from step #2 we clone extension repository into the` /tmp/projects/` folder and do checkout to the branch
	-  Compile extension (`yarn install` and `yarn compile`  --- I'm not sure if it will work for all extension)
	-  Next step is to prepare a devfile to create tested workspace. A template of this devfile is in `.ci/templates/extension-tests-devfile.yaml` and it looks like 

```yaml
apiVersion: 1.0.0
metadata:
  name: extension-test-sample
components:
  - mountSources: true
    memoryLimit: 3Gi
    type: dockerimage
    image: 'quay.io/eclipse/che-theia-dev:next'
    alias: che-dev
  - type: chePlugin
    reference: 'https://raw.githubusercontent.com/svor/che-vscode-extension-tests/main/meta.yaml'
  - type: chePlugin
    id: @
  - id: eclipse/che-theia/next
    memoryLimit: 1024Mi
    type: cheEditor
```

- When the devile is ready we can start the workspace. By using chectl we start the workspace and check when its status will be Running. When the workspce is ready, we copy extension's sources into /projects in theia-ide container. After that by using python .ci/tests-runner.py we run selenium script to open the workspace and run extension's tests.
- The last step is to check /projects/test.log file which should contain the result of test execution.

So, this step builds id of tested extension by analysing `che-theia-plugins.yaml`. If the changed plugin contains id we just take it, if not we build it from extension's `package.jon` as `publisher/name/latest`. When we have the id we need to replace @ with it in the devfile.

**How are tests running.**

The workspace contains `che-vscode-extension-tests` plug-in (it locates in local repository https://github.com/svor/che-vscode-extension-tests) it is included in the devfile:
```yaml
  - type: chePlugin
    reference: 'https://raw.githubusercontent.com/svor/che-vscode-extension-tests/main/meta.yaml'
```
When the workspace was started, this plugin automatically finds all *.test.js files in `/projects` folder inside the workspace and run them one by one via mocha framework.

For now tests are failed because each test needs to execute vscode.extensions.getExtension(id) to activate an extension and it gets `Cannot read property 'getExtension' of undefined`. There is an issue to investigate this problem: https://github.com/eclipse/che/issues/19172

### Screenshot/screencast of this PR
Some details of execution is here https://github.com/svor/che-plugin-registry/runs/1979560988?check_suite_focus=true

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/18219

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: .github/workflows/vscode-extension-tests-check.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      - name: Build image
+        if: env.EXTENSION_REPO != ''
+        run: |
+          export SKIP_FORMAT=true
+          export SKIP_LINT=true
+          export SKIP_TEST=true
+          docker pull quay.io/eclipse/che-plugin-registry:nightly
+          BUILDER=docker ./build.sh --tag vscode-extension 
+          # save locally built image
+          docker save -o docker-image.tar quay.io/eclipse/che-plugin-registry:vscode-extension
```
Comment (ericwill) [inline .github/workflows/vscode-extension-tests-check.yaml:43]: Let's try a better name than `vscode-extension`, maybe `vscode-extension-test` or something similar?

File: .ci/run-extension-tests.sh [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+
+set -e
+
+function installDeps() {
+    sudo pip install selenium
+    wget https://github.com/mozilla/geckodriver/releases/download/v0.29.0/geckodriver-v0.29.0-linux64.tar.gz
```
Comment (ericwill) [inline .ci/run-extension-tests.sh:18]: Please create an variable for the geckodriver version, and use that variable in the URL. This makes it easier to read and update.

File: .ci/run-extension-tests.sh [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+# Contributors:
+#   Red Hat, Inc. - initial API and implementation
+
+set -e
+
+function installDeps() {
+    sudo pip install selenium
```
Comment (ericwill) [inline .ci/run-extension-tests.sh:17]: It might be better to create a user level virtual environment.

File: .ci/tests-runner.py [hunk 4]
```diff
@@ -20,11 +20,15 @@
...
 options.headless = True
+options.add_argument('--disable-web-security')
+options.add_argument('--allow-running-insecure-content')
+options.add_argument('--ignore-certificate-errors')
 
 NEW_USER="admin"
 
 browser = webdriver.Firefox(options=options, executable_path="/usr/local/bin/geckodriver")
 wait = WebDriverWait(browser, 30)
+print (sys.argv[1])
```
Comment (ericwill) [inline .ci/tests-runner.py:31]: Looks like a print statement left over from development :smile:

File: .ci/templates/extension-tests-devfile.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+metadata:
+  name: extension-test-sample
+components:
+  - mountSources: true
+    memoryLimit: 3Gi
+    type: dockerimage
+    image: 'quay.io/eclipse/che-theia-dev:next'
+    alias: che-dev
+  - type: chePlugin
+    reference: 'https://raw.githubusercontent.com/svor/che-vscode-extension-tests/main/meta.yaml'
```
Comment (ericwill) [inline .ci/templates/extension-tests-devfile.yaml:11]: Is there a central place we can put this, i.e. not your fork?

**Review by ericwill — COMMENTED** [review]
Nice work @svor :tada:

I left some comments, see them inline

File: .ci/run-extension-tests.sh [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    export EXTENSION_PROJECT_NAME
+
+    mkdir -p /tmp/projects/"$EXTENSION_PROJECT_NAME"
+    git clone "${EXTENSION_REPO}" /tmp/projects/"$EXTENSION_PROJECT_NAME"
+    cd /tmp/projects/"$EXTENSION_PROJECT_NAME"
+    git checkout tags/"${EXTENSION_REVISION}"
+    git status
+}
+
+function prepareDevfile() {
```
Comment (benoitf) [inline .ci/run-extension-tests.sh:35]: let's move all the logic of this file to a github action

File: .ci/templates/extension-tests-devfile.yaml (hunk 5 above)
Comment (svor) [reply .ci/templates/extension-tests-devfile.yaml:11]: @ericwill I think we can move this extension into che-incubator, but when https://github.com/eclipse/che/issues/19172 will be resolved. I'm still working on this extension.

File: .ci/run-extension-tests.sh (hunk 6 above)
Comment (svor) [reply .ci/run-extension-tests.sh:35]: @benoitf do you mean to create more steps in the action and each step will execute some part of this script?

Comment (benoitf) [reply .ci/run-extension-tests.sh:35]: I was not clear enough

today you're using that script by doing:
```
name: Run VSCode extension's tests
        if: env.EXTENSION_REPO != ''
        run: |
          ./.ci/run-extension-tests.sh
```

but if you turn it to a github action, then we could just invoke it by calling

```
       if: env.EXTENSION_REPO != ''
        uses: che-incubator/vscode-extension-tests@next
```

(so plug-in registry is only a client of the action hosted elsewhere)

Comment (benoitf) [reply .ci/run-extension-tests.sh:35]: and with github actions it's easy to get unit tests / code coverage / etc.

**Comment by benoitf** [comment]
outdated

**Pull request closed without merging**